File: vuln-sandbox/typescript/cwe-89/sql-injection.ts
// VULNERABLE: SqlStringConcat — SQL string concatenation with user input (TypeScript)
// Rule: SqlStringConcat | CWE-89 | Severity: CRITICAL

import express, { Request, Response } from 'express';
import mysql2 from 'mysql2/promise';

const app = express();

async function getUser(req: Request, res: Response): Promise<void> {
  const db = await mysql2.createConnection({
    host: 'localhost',
    user: 'root',
    database: 'app',
  });

  const userId: string = req.params.id;

  // VULNERABLE: user-controlled input concatenated directly into SQL query
  // An attacker can pass: 1 OR 1=1 -- to dump all users
  const query: string = "SELECT * FROM users WHERE id = " + userId;
  const [rows] = await db.query(query);

  await db.end();
  res.json(rows);
}

app.get('/user/:id', getUser);
app.listen(3000);

File: vuln-sandbox/go/cwe-89/sql-injection.go
// VULNERABLE: go-sql-string-concat — SQL query built with string concatenation
// Rule: go-sql-string-concat | CWE-89 | Severity: HIGH

package main

import (
	"database/sql"
	"fmt"
	"net/http"

	_ "github.com/lib/pq"
)

var db *sql.DB

func getUserHandler(w http.ResponseWriter, r *http.Request) {
	userId := r.URL.Query().Get("id")

	// VULNERABLE: user-controlled input concatenated directly into SQL query
	// An attacker can pass: 1 OR 1=1 -- to dump all users
	query := "SELECT * FROM users WHERE id = " + userId
	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()
	fmt.Fprintf(w, "query executed")
}

func main() {
	var err error
	db, err = sql.Open("postgres", "postgres://localhost/app?sslmode=disable")
	if err != nil {
		panic(err)
	}
	http.HandleFunc("/user", getUserHandler)
	http.ListenAndServe(":8080", nil)
}
